AutomatedTesting/cmake/EngineFinder.cmake

```diff
@@ -47,46 +84,71 @@ Engine registration is required before configuring a project.
     endif()
 
-    math(EXPR engines_path_count "${engines_path_count}-1")
-    foreach(engine_path_index RANGE ${engines_path_count})
-        string(JSON engine_name ERROR_VARIABLE json_error MEMBER ${manifest_json} engines_path ${engine_path_index})
+    math(EXPR engines_count "${engines_count}-1")
+
+    set(most_compatible_engine_path "")
+    set(most_compatible_engine_version "")
+
+    foreach(array_index RANGE ${engines_count})
+        string(JSON engine_path ERROR_VARIABLE json_error GET ${manifest_json} engines "${array_index}")
         if(json_error)
-            message(FATAL_ERROR "Unable to read 'engines_path/${engine_path_index}' from '${manifest_path}'\nError: ${json_error}")
+            message(FATAL_ERROR "Unable to read 'engines/${array_index}' from '${manifest_path}'\nError: ${json_error}")
         endif()
 
-        if(LY_ENGINE_NAME_TO_USE STREQUAL engine_name)
-            string(JSON engine_path ERROR_VARIABLE json_error GET ${manifest_json} engines_path ${engine_name})
-            if(json_error)
-                message(FATAL_ERROR "Unable to read value from 'engines_path/${engine_name}'\nError: ${json_error}")
-            endif()
-
-            if(engine_path)
-                list(APPEND CMAKE_MODULE_PATH "${engine_path}/cmake")
-                return()
+        if(EXISTS ${engine_path}/engine.json)
+            o3de_engine_compatible(${engine_path} ${O3DE_ENGINE_NAME_TO_USE} ${O3DE_ENGINE_SPECIFIER_OP} ${O3DE_ENGINE_SPECIFIER_VERSION} is_compatible engine_version FALSE)
+            if(is_compatible)
+                if(NOT most_compatible_engine_path)
+                    set(most_compatible_engine_path ${engine_path})
+                    set(most_compatible_engine_version ${engine_version})
+                    if(ENGINE_FINDER_VERBOSE)
+                        message("-- Considering engine '${O3DE_ENGINE_NAME_TO_USE}' with version '${engine_version}' from '${engine_path}'")
+                    endif()
+                elseif(${engine_version} VERSION_GREATER ${most_compatible_engine_version})
+                    if(ENGINE_FINDER_VERBOSE)
+                        message("-- Considering engine '${O3DE_ENGINE_NAME_TO_USE}' with version '${engine_version}' from '${engine_path}' because it has a greater version number")
+                    endif()
+                    set(most_compatible_engine_path ${engine_path})
+                    set(most_compatible_engine_version ${engine_version})
+                elseif(ENGINE_FINDER_VERBOSE)
+                    message("-- Not considering engine '${O3DE_ENGINE_NAME_TO_USE}' with version '${engine_version}' from '${engine_path}' because it does not have a greater version number than the engine at '${most_compatible_engine_path}' (${most_compatible_engine_version})")
+                endif()
             endif()
+        else()
+            message(WARNING "${engine_path}/engine.json not found")
```

Comment:
The foreach loop looks like a good candidate for being put into a cmake function called `get_most_compatible_engine`